### ISSUE TITLE
zfs/zpool list on raidz2 pool seem to display faulty space values

### ISSUE THREAD

**Issue by rickatnight11** [zfsonlinux/zfs#3197]
I've just built a new Centos 7 box with 24 x 3TB drives, which I've added to a single raidz2 pool:

```
[root@stor01 media]# zpool status
  pool: array
 state: ONLINE
  scan: none requested
config:

        NAME                        STATE     READ WRITE CKSUM
        array                       ONLINE       0     0     0
          raidz2-0                  ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0
            wwn-0x50014eeOMITTED    ONLINE       0     0     0

errors: No known data errors
```

I've just filled it up with an existing dataset that I expected to fit on the array.  Taking a look at `zfs list` and `zpool list`, the values don't seem to match what I'd expect.

```
[root@stor01 media]# zpool list
NAME    SIZE  ALLOC   FREE    CAP  DEDUP  HEALTH  ALTROOT
array    65T  63.8T  1.22T    98%  1.00x  ONLINE  -
[root@stor01 media]# zfs list
NAME          USED  AVAIL  REFER  MOUNTPOINT
array        56.7T   188G   405K  /array
array/data    972G   188G   972G  /array/data
array/media  55.7T   188G  55.7T  /array/media
```

According to https://docs.oracle.com/cd/E26502_01/html/E29007/gbbti.html, `zpool list` should be showing `SIZE` as the raw disk space before the raid parity:
```
(24) * 3TiB = 72TiB = 73,728GiB
```
...yet it's showing what I'd expect the value to be usable space _after_ parity:
```
(24-2) * 3TiB = 66TiB = 67,584GiB
```
Furthermore, the base array's `USED` + `AVAIL` doesn't add up to the total `SIZE` of the zpool.  There appears to be about a `9TiB` discrepancy.

Am I missing something, or is there a bug in how these numbers are reported/calculated?

**Comment by DeHackEd** [comment]
Enormous RAID-Z arrays tend to have poorer space efficiency. You may be experiencing the penalty, especially if you have a lot of small files. Here 'small' is less than 96 kilobytes. Actually, since it's such a big number, even 128 block files will have unusually large overhead

This is unfortunately expected. `zfs list` has fudged the numbers to get something that seems sane.

**Comment by dirtyharrycallahan** [comment]
Is there an error in your calculation? Drives are typically sized in powers of 10. 24x 3 TB drives should work out to about 66 TiB.

**Comment by GeneralStumpf** [comment]
Several things to consider:
A) A 3TB drive (3*10^12 byte) is NOT a 3TiB (3*2^40 byte) drive.
B) HDD Vendors generally use the TB scale and add a few bytes, i.e. a 3TB labled HDD has 2,728+x TiB.
C) Are checksums enabled? If so, substract one 64th of the poolsize.
D) RAID Parity has to be substracted.
E) As mentioned by DeHackEd, the usable space also depends on file size (due to padding). For details on this have a look at Matt Ahrens' blog entry: http://blog.delphix.com/matt/2014/06/06/zfs-stripe-width/

tl;dr: the size zfs reports is correct.

**Comment by rickatnight11** [comment]
Ah, you're all absolutely right.  My math is bad, and I should feel bad.

```
3TB = ~2.7TiB
2.7TiB * 24 = 64.8TiB
```
(http://www.wolframalpha.com/input/?i=24+*+%282.7+tebibytes%29)

That means the reported `zpool list` size is correct.

Taking raidz2 into account, that least me with a theoretical:

```
2.7TiB * 22 = 59.4TiB
```

That leaves about a `2TiB` discrepancy, `~1TiB` of which can be attributed to checksums, and the rest to padding.

Does that all sound right?

**Comment by GregorKopka** [comment]
@rickatnight11 unrelated comment:
24*3TB drives in a single raidz2 is - according to some discussions on the mailing list - a desaster waiting to happen: long resilver times with high stress to the surviving drives -> additional drive failures -> pool loss

**Comment by rickatnight11** [comment]
Totally agree.  All data is backed up regularly to tape and easily recoverable.  ZFS ends up being the best compromise between convenience and performance in this case.  Closing out.  Thanks, everyone!
